### ISSUE TITLE
PyICU fails to compile with ICU 59.1

### ISSUE THREAD

**Issue by erosennin** [ovalhub/pyicu#75]
Both 2.0.2 and the current master fail to compile with ICU 59.1:

```
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-2.7.13=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c bases.cpp -o build/temp.linux-x86_64-2.7/bases.o -DPYICU_VER="2.0.2" -std=c++11 -I/usr/local/include
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
In file included from bases.cpp:30:0:
bases.cpp: In function 'void _init_bases(PyObject*)':
bases.cpp:2368:27: error: 'U_TITLECASE_NO_BREAK_ADJUSTMENT' was not declared in this scope
     INSTALL_MODULE_INT(m, U_TITLECASE_NO_BREAK_ADJUSTMENT);
                           ^
macros.h:371:44: note: in definition of macro 'INSTALL_MODULE_INT'
     PyModule_AddIntConstant(module, #name, name);
                                            ^~~~
bases.cpp:2369:27: error: 'U_TITLECASE_NO_LOWERCASE' was not declared in this scope
     INSTALL_MODULE_INT(m, U_TITLECASE_NO_LOWERCASE);
                           ^
macros.h:371:44: note: in definition of macro 'INSTALL_MODULE_INT'
     PyModule_AddIntConstant(module, #name, name);
                                            ^~~~
error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

The error can be reproduced with this Dockerfile:

```
FROM debian:stable

RUN apt-get -y update && apt-get -y install git wget build-essential python-dev

RUN cd /tmp
RUN wget http://download.icu-project.org/files/icu4c/59.1/icu4c-59_1-src.tgz
RUN tar xf icu4c-59_1-src.tgz
RUN cd icu/source && ./configure && make -j3 && make install

RUN cd /tmp
RUN git clone https://github.com/ovalhub/pyicu/
RUN cd pyicu && python setup.py build
```
```

**Comment by ovalhub** [comment]
On Wed, 24 Jan 2018, Andrey Golovizin wrote:

> Both 2.0.2 and the current master fail to compile with ICU 59.1:
>
> ```
> x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-2.7.13=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c bases.cpp -o build/temp.linux-x86_64-2.7/bases.o -DPYICU_VER="2.0.2" -std=c++11 -I/usr/local/include
> cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
> In file included from bases.cpp:30:0:
> bases.cpp: In function 'void _init_bases(PyObject*)':
> bases.cpp:2368:27: error: 'U_TITLECASE_NO_BREAK_ADJUSTMENT' was not declared in this scope
>     INSTALL_MODULE_INT(m, U_TITLECASE_NO_BREAK_ADJUSTMENT);

This error is strange as U_TITLECASE_NO_BREAK_ADJUSTMENT is available since 
ICU 3.8.
http://icu-project.org/apiref/icu4c/stringoptions_8h.html#afc65fa226cac9b8eeef0e877b8a7744e

> macros.h:371:44: note: in definition of macro 'INSTALL_MODULE_INT'
>     PyModule_AddIntConstant(module, #name, name);
>                                            ^~~~
> bases.cpp:2369:27: error: 'U_TITLECASE_NO_LOWERCASE' was not declared in this scope
>     INSTALL_MODULE_INT(m, U_TITLECASE_NO_LOWERCASE);

Same comment, U_TITLECASE_NO_LOWERCASE is available since ICU 3.8.

Other people that are building PyICU with ICU 53 (since that is the 
installed version in the current Ubuntu LTS it seems) didn't report errors 
about these ints.

Could it be that you're not including the correct header files ?

Andi..

**Comment by ovalhub** [comment]
It could also be that the stringoptions.h file needs to be included 
explcitely. I know it was introduced at some point and constants moved into 
it.
I'm asking the ICU people when that was so that I can add the proper 
include.

Andi..

**Comment by ovalhub** [comment]
The stringoptions.h file was introduced in ICU 60, not affecting you.
Still, I also added an include for casemap.h, just in case.

Please, try HEAD again and let me know if this solves the problem.
If not, please add an assert someplace that checks that
   U_ICU_VERSION_HEX == VERSION_HEX(59, 1, 0)
in your setup.

Andi..

**Comment by erosennin** [comment]
OK, after some archaeology, it looks like the correct incude is [ucasemap.h](http://bugs.icu-project.org/trac/browser/trunk/icu4c/source/common/unicode/ucasemap.h). :)

1. The file exists since at least ICU 3.4 and claims to define `U_TITLECASE_NO_LOWERCASE` and stuff [since ICU 3.8](http://bugs.icu-project.org/trac/browser/trunk/icu4c/source/common/unicode/ucasemap.h?rev=40161#L147).
2. It used to be included by [unistr.h](http://bugs.icu-project.org/trac/browser/trunk/icu4c/source/common/unicode/unistr.h?rev=39583#L36) in older versions.  PyICU includes `unistr.h` via [common.h](https://github.com/ovalhub/pyicu/blob/f81fa3ff33ee9a5d143d8b83b48c14ffa25e1251/common.h#L106) so everything used to compile fine until the incude was [removed](http://bugs.icu-project.org/trac/changeset/39684/trunk/icu4c/source/common/unicode/unistr.h#file3) from `unistr.h` somewhere before ICU 59.
  
2. In ICU 60, `U_TITLECASE_NO_LOWERCASE` and friends were [moved](http://bugs.icu-project.org/trac/changeset/40162#file4) to [stringoptions.h](http://bugs.icu-project.org/trac/browser/tags/release-60-2/icu4c/source/common/unicode/stringoptions.h#L72) which is still included by [ucasemap.h](http://bugs.icu-project.org/trac/browser/tags/release-60-2/icu4c/source/common/unicode/ucasemap.h#L26).

TL;DR adding `#include <unicode/ucasemap.h>` fixes compilation with ICU 59.1 and should not break anything with other ICU versions. :)

**Comment by ovalhub** [comment]
Thank you for elucidating this, I fixed the bad include I added earlier in HEAD.
ucasemap.h is now included for all versions.
